For #12287: Add Synced Tabs to Tabs Tray

jonalmeida opened the pull request:
The feature is put behind a feature flag from the Secret Settings. Might be easier to review per commit.

Abstract [design link](https://share.goabstract.com/0bae23a7-b55b-43f9-a57a-1fdb7b3596fc?mode=build&sha=a813facef071a6c43e3d40d20afa4aaf4f75e269) to verify against.

![out](https://user-images.githubusercontent.com/1370580/90475971-8ff50e80-e0f6-11ea-953f-3ef0849169ec.png)


### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

jonalmeida commented:
The parts I'm most unsure of are the navigation graph additions.

ekager commented on app/src/main/res/layout/view_synced_tabs_title.xml:
```diff
@@ -0,0 +1,29 @@
+        android:text="@string/synced_tabs" />
+
+    <ImageView
+        android:id="@+id/refresh_icon"
```
nit - let's give this a `selectableItemBackground` so it visually clicks

ekager replied:
Same with the synced tabs items - they should have a `selectableItemBackground` 
`

jonalmeida replied:
> Same with the synced tabs items - they should have a `selectableItemBackground`

I already see that the synced tab items have the ripple animation when long pressed. Are you referring to something else?

ekager replied:
Okay I see the ripple it's just verrrryyy faint on dark mode. Also the padding/margin is uneven when you click it - let's fix that then 😅

jonalmeida replied:
Yeah! I actually noticed and fixed that because of this comment and then wasn't sure if _that_ was what you were referring to. 😁

ekager replied:
Brain waves 🧠 〰️

jonalmeida commented:
Thanks for the review; super helpful!

>  1. Do we want this to show in the private tab in the tabs tray?

We don't see it in private tabs, I'll leave it up to @topotropic to comment (also consider the separate tab idea that Sorin mentioned).

>  2. We probably want to hide this in multiselect mode like we do for the save to collection button because otherwise it looks a bit strange?

Fixed! Check out the last commit that addresses the review comments.

>  3. When opening the tabs tray with synced tabs enabled and 1 open tab, it seems to jump to the synced tabs section and opens at an awkward point in the list.
>  4. Sometimes (can't repro 100%) it even opens further down on the synced tabs list and I would have to scroll _way_ up to see open tabs. When I swipe up to try to scroll, it closes. Got a video from after it opened.

These two are related and are a side-effect to how the tabs tray scrolls to the selected tab. The hacky approach that we currently have (calculate the index based on the adapter size), does not consider that the adapter is loaded async and we need to scroll when all the adapters are complete.

I've filed https://github.com/mozilla-mobile/fenix/issues/13905 to address this and added it as a blocker for removing the feature flag.

> Moving this into it's own Tabs Tray Tab (phew mouthful) would likely solve all these problems, so if we decide to move that way (via this comment [#12287 (comment)](https://github.com/mozilla-mobile/fenix/issues/12287#issuecomment-675285951) ) we can probably disregard these issues.

Yes! I'd really like us to look into this but I don't think it'll block us from landing this change.
